---
title: "Vegetation Data: dietary analysis"
author: "Magali Corti"
date: "5/16/2023"
output: 
  html_document: 
    toc: yes
---

This is the script that I made for the filtering of the raw data obtained from the bioinformatic analysis. The outputs of the dietary analysis are divided into several dataframes, F and M are two different libraries, while EU, BR, FU, and GH are the four primers that have been used for the DNA metabarcoding analysis (corresponding respectively to Eukaryota, Bryophytes, Fungi and Vascular Plants).
Another dataset is the one obtained from the REININ project, characterizing the the diet of reindeer in Finnmark (the same study area of the previuos data), in this case the primers used were targeting Vascular plants (GH), and Eukaryotes (EUKA).
Here I am going to filter and merge the raw datasets, with the goal of having a single dataset for each primer set, I will select the samples used in my study, and crate the dietary data that I will use to compare with the microbiome data.

# Data filtering

Loading packages
```{r message=FALSE, warning=FALSE, results='hide'}
pack <- c("tidyverse","janitor","vegan","phyloseq","decontam","devtools","Rtools")
sapply(pack, require, character.only = TRUE)
```


## Eukariotic data

#### Importing data
```{r}
# importing .tab files (tab separated files)
F_EU <- read.table(file = "data/F_EU.tab", header = T, sep = "\t")
M_EU <- read.table(file = "data/M_EU.tab", header = T, sep = "\t")
Finn_EU <- read.table(file = "data/Finn_EU.tab", header = T, sep = "\t")
```

#### Erasing "obiclean" columns
```{r}
# selecting columns not starting with "obi" 
F_EU <- F_EU %>% 
  dplyr::select(-(starts_with('obi')))
M_EU <- M_EU %>% 
  dplyr::select(-(starts_with('obi')))
Finn_EU <- Finn_EU %>% 
  dplyr::select(-(starts_with('obi')))
```

#### Excluding MOTUs defined only as Eukaryota (superkingdom level) 
```{r}
# subsetting dataframe: all rows which scientific name is different than Eukaryota
F_EU <- F_EU %>% 
  subset(scientific_name!="Eukaryota")
M_EU <- M_EU %>% 
  subset(scientific_name!="Eukaryota")
Finn_EU <- Finn_EU %>% 
  subset(scientific_name!="Eukaryota")
```

#### Filtering data that has less than 97% identity with the Eukariotic reference dataset
97% is the minimal, and one of the thresholds that are most used in the analysis of metabarcoding outputs.
```{r}
# subsetting dataframe: all rows which best_identity.db_EUKA is greater than 0.97
F_EU <- F_EU %>% 
  subset(best_identity.db_EUKA>0.97)
M_EU <- M_EU %>% 
  subset(best_identity.db_EUKA>0.97)
Finn_EU <- Finn_EU %>% 
  subset(best_identity.db_EUKA>0.97)
```

#### Erasing "strange" data
In our data, even after the previous data filtering during bioinformatic analysis, we have some probable mis-assignement or some contamination from other sources of DNA. Since our faeces samples come from herbivore species we assume that Gallus gallus shouldn't be on the list of dietary items, nor other mammals (even though reindeer have been occasionally witnessed feeding on lemmings during the winter).
```{r}
# subsetting dataframe: 
# excluding data that have rank of kingdom
# not that much informative
F_EU <- F_EU %>% 
  subset(rank!="kingdom")
M_EU <- M_EU %>% 
  subset(rank!="kingdom")
Finn_EU <- Finn_EU %>% 
  subset(rank!="kingdom")
```


#### Subsetting and merging datasets for Eukaryota
The dataset that we have is quite big, while for our project we used only a total of 40 samples, here we select the one that we will be using.
```{r}
# selecting only column from M_EU dataset with sequence and our samples
M_EU_join <- M_EU %>% select(sequence, sample.REINS1P2Q2_S413_R2, sample.REINS1P2Q2_S413_R3, sample.REINS1P2Q3_S415_R2, sample.REINS1P2Q3_S415_R3)

# selecting only column from F_EU dataset with sequence, scientific name and our samples
F_EU_join <- F_EU %>% select(sequence, scientific_name, sample.MS1P1Q2_SH135_R1, sample.MS1P1Q2_SH135_R2, sample.MS1P1Q2_SH135_R3, sample.MS1P2Q2_SH164_R2, sample.MS1P2Q2_SH164_R3 , sample.MS1P2Q4_SH134_R1, sample.MS1P2Q4_SH134_R2, sample.MS1P2Q4_SH134_R3, sample.MS1P2Q4_SH136_R1, sample.MS1P2Q4_SH136_R2, sample.MS1P2Q4_SH136_R3, sample.MS1P3Q1_SH132_R1, sample.MS1P3Q1_SH132_R2, sample.MS1P3Q1_SH132_R3, sample.MS1P3Q1_SH133_R1, sample.MS1P3Q1_SH133_R2, sample.MS1P3Q1_SH133_R3, sample.MS2P1Q1_SH195_R1, sample.MS2P1Q1_SH195_R2, sample.MS2P1Q1_SH195_R3, sample.MS2P2Q1_SH196_R1, sample.MS2P2Q1_SH196_R2, sample.MS2P2Q1_SH196_R3, sample.HS1P1Q4_SH137_R1, sample.HS1P1Q4_SH137_R2, sample.HS1P1Q4_SH137_R3, sample.HS2P1Q3_SH193_R1, sample.HS2P1Q3_SH193_R2, sample.HS2P1Q3_SH193_R3, sample.HS2P2Q2_SH194_R1, sample.HS2P2Q2_SH194_R2, sample.HS2P2Q2_SH194_R3, sample.HS2P2Q3_SH138_R1, sample.HS2P2Q3_SH138_R2, sample.HS2P2Q3_SH138_R3, sample.HS2P2Q3_SH190_R1, sample.HS2P2Q3_SH190_R2, sample.HS2P2Q3_SH190_R3, sample.HS2P2Q3_SH192_R1, sample.HS2P2Q3_SH192_R2, sample.HS2P2Q3_SH192_R3, sample.HS2P3Q3_SH140_R1, sample.HS2P3Q3_SH140_R2, sample.HS2P3Q3_SH140_R3, sample.HS2P3Q4_SH139_R1, sample.HS2P3Q4_SH139_R2, sample.HS2P3Q4_SH139_R3, sample.PS2P1Q3_SH80_R1, sample.PS2P1Q3_SH80_R2, sample.PS2P1Q3_SH80_R3, sample.PS2P2Q2_SH100_R1, sample.PS2P2Q2_SH100_R2, sample.PS2P2Q2_SH100_R3, sample.PS2P2Q2_SH186_R1, sample.PS2P2Q2_SH186_R2, sample.PS2P2Q2_SH186_R3, sample.PS2P2Q2_SH81_R1, sample.PS2P2Q2_SH81_R2, sample.PS2P2Q2_SH81_R3, sample.PS2P2Q3_SH115_R1, sample.PS2P2Q3_SH115_R2, sample.PS2P2Q3_SH115_R3, sample.PS2P2Q4_SH113_R1, sample.PS2P2Q4_SH113_R2, sample.PS2P2Q4_SH113_R3, sample.PS2P2Q4_SH98_R1, sample.PS2P2Q4_SH98_R2, sample.PS2P2Q4_SH98_R3, sample.PS2P3Q3_SH99_R1, sample.PS2P3Q3_SH99_R2, sample.PS2P3Q3_SH99_R3, sample.RS1P2Q2_SH124_R1,  sample.RS1P2Q2_SH124_R2,  sample.RS1P2Q2_SH124_R3, sample.RS1P2Q1_SH106_R1, sample.RS1P2Q1_SH106_R2, sample.RS1P2Q1_SH106_R3, sample.RS2P1Q1_SH129_R1, sample.RS2P1Q1_SH129_R2, sample.RS2P1Q1_SH129_R3, sample.RS2P1Q2_SH93_R1, sample.RS2P1Q2_SH93_R2, sample.RS2P1Q2_SH93_R3, sample.RS2P2Q3_SH127_R1, sample.RS2P2Q3_SH127_R2, sample.RS2P2Q3_SH127_R3, sample.RS2P2Q4_SH130_R1, sample.RS2P2Q4_SH130_R2, sample.RS2P2Q4_SH130_R3, sample.RS2P2Q4_SH172_R1, sample.RS2P2Q4_SH172_R2, sample.RS2P2Q4_SH172_R3, sample.RS2P2Q4_SH176_R1, sample.RS2P2Q4_SH176_R2, sample.RS2P2Q4_SH176_R3, sample.REINS1P1Q5_S412_R1, sample.REINS1P1Q5_S412_R2, sample.REINS1P1Q5_S412_R3, sample.REINS1P2Q45_S414_R1, sample.REINS1P2Q45_S414_R3, sample.NC2_FIN_R1, sample.NC2_FIN_R2, sample.NC2_FIN_R3, sample.NC3_FIN_R1, sample.NC3_FIN_R2, sample.NC2_FIN_R3, sample.NC4_FIN_R1, sample.NC4_FIN_R2, sample.NC4_FIN_R3, sample.NC5_FIN_R1, sample.NC5_FIN_R2, sample.NC5_FIN_R3, sample.NC6_FIN_R1, sample.NC6_FIN_R2, sample.NC6_FIN_R3, sample.NC7_FIN_R1, sample.NC7_FIN_R2, sample.NC7_FIN_R3, sample.NC_FIN_R1, sample.NC_FIN_R2, sample.NC_FIN_R3, sample.PCRNC2_FIN_R1, sample.PCRNC2_FIN_R2, sample.PCRNC2_FIN_R3, sample.PCRNC3_FIN_R2, sample.PCRNC3_FIN_R3, sample.PCRNC4_FIN_R3, sample.NC_FIN_R1, sample.NC_FIN_R2, sample.NC_FIN_R3, sample.PCRNC_FIN_R1, sample.PCRNC_FIN_R2, sample.PCRNC_FIN_R3)

# selecting only column from Finn_EU dataset with sequence and our samples
Finn_EU_join <- Finn_EU %>% select(sequence, sample.S1562_R1, sample.S1562_R2, sample.S1562_R3, sample.S1564_R1, sample.S1564_R2, sample.S1564_R3, sample.S1568_R1, sample.S1568_R2, sample.S1568_R3, sample.S1572_R1, sample.S1572_R2, sample.S1572_R3)

# merging F_EU and M_EU in one single FM_EU dataset
FM_EU <- full_join(F_EU_join, M_EU_join, by = "sequence", copy = FALSE, keep = NULL)
FMFinn_EU <- full_join(FM_EU, Finn_EU_join, by = "sequence", copy = FALSE, keep = NULL)
```


#### Dereplication of MOTUs with the same scientific name
It can happen that different sequences are assigned to the same scientific name, in order to do a more sound and easy analysis is good to dereplicate these data, resulting a dataframe with unique names and the total number of reads for each one of them.
```{r results='hide'}
# Check if we have duplicated scientific_name
table(duplicated(FMFinn_EU$scientific_name))

# 17170 taxa names are duplicates
  
# Dereplication by scientific_name
FMFinn_EU_derep <- FMFinn_EU %>% 
  group_by(scientific_name) %>% 
  summarise_if(is.numeric, sum)

# remove last row from data frame (NAs)
FMFinn_EU_derep <- FMFinn_EU_derep %>% filter(row_number() <= n()-1)

# replace all NA values with zero
FMFinn_EU_derep <- FMFinn_EU_derep %>% replace(is.na(.), 0)
```


#### Resolving PCR replicates
Each sample have been amplified thrice, in order to have three PCR replicates, hence we need to dereplicate the results and find the total number of reads for each sample.
On the other hand is very useful to create a matrix of the frequency of taxa detected across 3 samples (especially when there are many replicates).
```{r}
# We exclude first column (scientific_name), transpose and convert the data matrix in dataframe
FMFinn_EU_derep_R <- data.frame(t(FMFinn_EU_derep[,-1]))
# AND rename the table by scientific name
names(FMFinn_EU_derep_R) <- FMFinn_EU_derep$scientific_name

# make sample names one of the columns of the data
FMFinn_EU_derep_R <- FMFinn_EU_derep_R %>% 
  rownames_to_column("Sequencing_ID")

# Prepare an id that represents all three replicates (by same name) and place that after "Sequencing_ID"
FMFinn_EU_derep_R <- FMFinn_EU_derep_R %>% 
  add_column(Sample_ID = str_replace_all(FMFinn_EU_derep_R$Sequencing_ID, "_R1|_R2|_R3", ""), .after = "Sequencing_ID")

# get read sum of all three replicates
FMFinn_EU_clean_Rsum <- FMFinn_EU_derep_R %>% 
  dplyr::select(-Sequencing_ID) %>% 
  group_by(Sample_ID) %>% 
  summarise_if(is.numeric, sum)
# We have the read data in required format (Sample X Species)

# Convert un-merged read data to 0 or 1, remove first two columns
#FMFinn_EU_derep_PCR <- data.frame(ifelse (FMFinn_EU_derep_R[,-c(1:2)]>0, 1, 0))

# Combine with the first two columns of the previous data
#FMFinn_EU_derep_PCR <- cbind(FMFinn_EU_derep_R[,c(1:2)], FMFinn_EU_derep_PCR)

# Follow the steps used while generating read sums
#FMFinn_EU_clean_PCR <- FMFinn_EU_derep_PCR %>% 
#  dplyr::select(-Sequencing_ID) %>% 
#  group_by(Sample_ID) %>% 
#  summarise_if(is.numeric, sum)
#FMFinn_EU_clean_PCR <- FMFinn_EU_clean_PCR[-1,]
# We have the frequency data in required format (Sample X Species)
```


#### Identifying contaminants in marker-gene data

Creating FMFinn_EU metadata dataframe with a Sample_or_Control column for running decontam function
```{r}
FMFinn_EU_meta <- as.data.frame(FMFinn_EU_clean_Rsum$Sample_ID)
names(FMFinn_EU_meta) <- "Sample_ID"

FMFinn_EU_meta$Sample_or_Control <- "True Sample"
FMFinn_EU_meta$Sample_or_Control[grepl("NC", FMFinn_EU_meta$Sample_ID)] <- "Control Sample"

FMFinn_EU_meta <- FMFinn_EU_meta[-1,]
```

Creating phyloseq object
```{r}
EU_to_ps <- as.data.frame(FMFinn_EU_clean_Rsum[-1,])

# first column to rowname
row.names(EU_to_ps) <- EU_to_ps$Sample_ID
EU_to_ps <- EU_to_ps[,-1]

ps_EU <- phyloseq(otu_table(EU_to_ps, taxa_are_rows=T))
```

Identify Contaminants - with Prevalence method
```{r}
neg <- FMFinn_EU_meta$Sample_or_Control == 'Control Sample' # True if a negative control sample

# Prevalence-based contaminant classification
contam <- isContaminant(as.matrix(ps_EU), neg=neg, threshold=0.1, detailed=TRUE, normalize=TRUE, method='prevalence')

table(contam$contaminant)

#contam
```

Erasing from EU dataset the taxa identifed as contaminants
```{r}
# right format for df to merge
tFMFinn_EU_clean <- as.data.frame(t(FMFinn_EU_clean_Rsum))
tFMFinn_EU_clean <- tFMFinn_EU_clean %>% row_to_names(row_number = 1)
tFMFinn_EU_clean <- tibble::rownames_to_column(tFMFinn_EU_clean, "Taxa")

contam <- tibble::rownames_to_column(contam, "Taxa")

contaminant <- contam %>%
   dplyr::select(Taxa, contaminant)

# merge contaminant column with EU dataset
FMFinn_EU_contam <- left_join(tFMFinn_EU_clean, contaminant, by = "Taxa")

# subset only non contaminants
FMFinn_EU_decontam <- FMFinn_EU_contam %>%
  subset(contaminant == "FALSE") %>%
  dplyr::select(-contaminant)
```


#### Removing rare reads
We need to decide if for the purposes of our project we want to select a treshold which is a percentage of the dataset (eg. 1 per thousand most abundant MOTUs) or an arbitrary number of MOTUs considered as predominant in the diet (eg. 50 most abundant MOTUs).
```{r}
# Erasing negative controls
FMFinn_EU_decontam <- FMFinn_EU_decontam %>% 
       select(-contains('NC'))

# convert from character to numeric
Taxa <- as.data.frame(FMFinn_EU_decontam$Taxa)
FMFinn_EU_decontam <- as.data.frame(apply(FMFinn_EU_decontam[,-1], 2, as.numeric))
FMFinn_EU_decontam <- cbind(Taxa, FMFinn_EU_decontam)
names(FMFinn_EU_decontam)[1] = "Taxa"

# creating a count column with the sum of the readings
rf <- FMFinn_EU_decontam %>% rowwise(Taxa)
FMFinn_EU_decontam <- rf %>% mutate(count = sum(c_across(where(is.numeric))))

# ordering data by decreasing number of counts
order(FMFinn_EU_decontam$count, decreasing = T)
FMFinn_EU_decontam <- FMFinn_EU_decontam[order(FMFinn_EU_decontam$count, decreasing = T),]

# keeping only first 50 most abundant MOTUs
FMFinn_EU_50 <- FMFinn_EU_decontam[1:50, ]
```


#### Exporting dataframes

Changing format
```{r}
# transposing
tFMFinn_EU_50 <- as.data.frame(t(FMFinn_EU_50))
tFMFinn_EU_50 <- tFMFinn_EU_50 %>% 
  row_to_names(row_number = 1) 
FMFinn_EU_final <- tFMFinn_EU_50[-41,] 
#FMFinn_EU_final <- tibble::rownames_to_column(FMFinn_EU_final, "Sample_ID")
```

Exporting dataframes
```{r}
# exporting the dataframe of our samples in csv format
write.csv(FMFinn_EU_final, "outputs/FMFinn_EU_final.csv", row.names = T)
write.csv(FMFinn_EU_50, "outputs/FMFinn_EU_50.csv", row.names = F)
```



## Bryophyte data

#### Importing data
```{r}
# importing .tab files (tab separated files)
F_BR <- read.table(file = "data/F_BR.tab", header = T, sep = "\t")
M_BR <- read.table(file = "data/M_BR.tab", header = T, sep = "\t")
Finn_BR <- read.table(file = "data/Finn_BR.tab", header = T, sep = "\t")
```

#### Erasing "obiclean" columns
```{r}
# selecting columns not starting with "obi" 
F_BR <- F_BR %>% 
  select(-(starts_with('obi')))
M_BR <- M_BR %>% 
  select(-(starts_with('obi')))
Finn_BR <- Finn_BR %>% 
  dplyr::select(-(starts_with('obi')))
```

#### Excluding MOTUs defined only as Eukaryota (superkingdom level) 
```{r}
# subsetting dataframe: all rows which scintific name is different than Eukaryota
F_BR <- F_BR %>% 
  subset(scientific_name!="Eukaryota")
M_BR <- M_BR %>% 
  subset(scientific_name!="Eukaryota")
Finn_BR <- Finn_BR %>% 
  subset(scientific_name!="Eukaryota")
```

#### Filtering data that has less than 97% identity with the Bryophyte reference dataset
```{r}
# subsetting dataframe: all rows which best_identity.db_BRYO is greater than 0.97
F_BR <- F_BR %>% 
  subset(best_identity.db_BRYO>0.97)
M_BR <- M_BR %>% 
  subset(best_identity.db_BRYO>0.97)
Finn_BR <- Finn_BR %>% 
  subset(best_identity.db_BRYO>0.97)
```

#### Erasing not informative data
```{r}
# erasing data that have rank of kingdom
F_BR <- F_BR %>% 
  subset(rank!="kingdom")
M_BR <- M_BR %>% 
  subset(rank!="kingdom")
Finn_BR <- Finn_BR %>% 
  subset(rank!="kingdom")
```

#### Subsetting and merging datasets for Bryophytes
```{r}
# selecting only column from M_BR dataset with sequence and our samples
M_BR_join <- M_BR %>% select(sequence, sample.REINS1P2Q2_S413_R1, sample.REINS1P2Q2_S413_R3, sample.REINS1P2Q3_S415_R1, sample.REINS1P2Q3_S415_R3, sample.EXTRNC_168_R1, sample.EXTRNC_168_R3, sample.NC210618_R1, sample.NC210618_R3, sample.NC_MIX_137_R1, sample.NC_MIX_R1, sample.NC_MIX_137_R2, sample.NC_MIX_R3, sample.PCRNC_MIX_1_R1, sample.PCRNC_MIX_2_R1, sample.PCRNC_MIX_3_R1, sample.PCRNC_MIX_4_R1)

# selecting only column from F_BR dataset with sequence, scientific name and our samples
F_BR_join <- F_BR %>% select(sequence, scientific_name, sample.MS1P1Q2_SH135_R1, sample.MS1P1Q2_SH135_R2, sample.MS1P1Q2_SH135_R3, sample.MS1P2Q2_SH164_R2, sample.MS1P2Q2_SH164_R3 , sample.MS1P2Q4_SH134_R1, sample.MS1P2Q4_SH134_R2, sample.MS1P2Q4_SH134_R3, sample.MS1P2Q4_SH136_R1, sample.MS1P2Q4_SH136_R2, sample.MS1P2Q4_SH136_R3, sample.MS1P3Q1_SH132_R1, sample.MS1P3Q1_SH132_R2, sample.MS1P3Q1_SH132_R3, sample.MS1P3Q1_SH133_R1, sample.MS1P3Q1_SH133_R2, sample.MS1P3Q1_SH133_R3, sample.MS2P1Q1_SH195_R1, sample.MS2P1Q1_SH195_R2, sample.MS2P1Q1_SH195_R3, sample.MS2P2Q1_SH196_R1, sample.MS2P2Q1_SH196_R2, sample.MS2P2Q1_SH196_R3, sample.HS1P1Q4_SH137_R1, sample.HS1P1Q4_SH137_R2, sample.HS1P1Q4_SH137_R3, sample.HS2P1Q3_SH193_R1, sample.HS2P1Q3_SH193_R2, sample.HS2P1Q3_SH193_R3, sample.HS2P2Q2_SH194_R1, sample.HS2P2Q2_SH194_R2, sample.HS2P2Q2_SH194_R3, sample.HS2P2Q3_SH138_R1, sample.HS2P2Q3_SH138_R2, sample.HS2P2Q3_SH138_R3, sample.HS2P2Q3_SH190_R1, sample.HS2P2Q3_SH190_R2, sample.HS2P2Q3_SH190_R3, sample.HS2P2Q3_SH192_R1, sample.HS2P2Q3_SH192_R2, sample.HS2P2Q3_SH192_R3, sample.HS2P3Q3_SH140_R1, sample.HS2P3Q3_SH140_R2, sample.HS2P3Q3_SH140_R3, sample.HS2P3Q4_SH139_R1, sample.HS2P3Q4_SH139_R2, sample.HS2P3Q4_SH139_R3, sample.PS2P1Q3_SH80_R1, sample.PS2P1Q3_SH80_R2, sample.PS2P1Q3_SH80_R3, sample.PS2P2Q2_SH100_R1, sample.PS2P2Q2_SH100_R2, sample.PS2P2Q2_SH100_R3, sample.PS2P2Q2_SH186_R1, sample.PS2P2Q2_SH186_R2, sample.PS2P2Q2_SH186_R3, sample.PS2P2Q2_SH81_R1, sample.PS2P2Q2_SH81_R2, sample.PS2P2Q2_SH81_R3, sample.PS2P2Q3_SH115_R1, sample.PS2P2Q3_SH115_R2, sample.PS2P2Q3_SH115_R3, sample.PS2P2Q4_SH113_R1, sample.PS2P2Q4_SH113_R2, sample.PS2P2Q4_SH113_R3, sample.PS2P2Q4_SH98_R1, sample.PS2P2Q4_SH98_R2, sample.PS2P2Q4_SH98_R3, sample.PS2P3Q3_SH99_R1, sample.PS2P3Q3_SH99_R2, sample.PS2P3Q3_SH99_R3, sample.RS1P2Q2_SH124_R1,  sample.RS1P2Q2_SH124_R2,  sample.RS1P2Q2_SH124_R3, sample.RS1P2Q1_SH106_R1, sample.RS1P2Q1_SH106_R2, sample.RS1P2Q1_SH106_R3, sample.RS2P1Q1_SH129_R1, sample.RS2P1Q1_SH129_R2, sample.RS2P1Q1_SH129_R3, sample.RS2P1Q2_SH93_R1, sample.RS2P1Q2_SH93_R2, sample.RS2P1Q2_SH93_R3, sample.RS2P2Q3_SH127_R1, sample.RS2P2Q3_SH127_R2, sample.RS2P2Q3_SH127_R3, sample.RS2P2Q4_SH130_R1, sample.RS2P2Q4_SH130_R2, sample.RS2P2Q4_SH130_R3, sample.RS2P2Q4_SH172_R1, sample.RS2P2Q4_SH172_R2, sample.RS2P2Q4_SH172_R3, sample.RS2P2Q4_SH176_R1, sample.RS2P2Q4_SH176_R2, sample.RS2P2Q4_SH176_R3, sample.REINS1P1Q5_S412_R1, sample.REINS1P1Q5_S412_R2, sample.REINS1P1Q5_S412_R3, sample.REINS1P2Q45_S414_R1, sample.REINS1P2Q45_S414_R3, sample.NC2_FIN_R1, sample.NC_FIN_R3, sample.OANC2_R1)

# selecting only column from Finn_BR dataset with sequence and our samples
Finn_BR_join <- Finn_BR %>% select(sequence, sample.S1562_R1, sample.S1562_R2, sample.S1562_R3, sample.S1564_R1, sample.S1564_R2, sample.S1564_R3, sample.S1568_R1, sample.S1568_R2, sample.S1568_R3, sample.S1572_R1, sample.S1572_R2, sample.S1572_R3, sample.DNA_NC_F2_R1, sample.DNA_NC_F2_R3, sample.DNANCF11_R1, sample.DNANCF11_R2, sample.DNANCF11_R3, sample.DNANCF4_R1, sample.DNANCF4_R1, sample.DNANCF4_R2, sample.DNANCF4_R3, sample.PCRNCF4_R1, sample.PCRNCF4_R3)

# merging F_BR and M_BR in one single FM_BR dataset
FM_BR <- full_join(F_BR_join, M_BR_join, by = "sequence", copy = FALSE, keep = NULL)
FM_BR <- full_join(FM_BR, Finn_BR_join, by = "sequence", copy = FALSE, keep = NULL)
```

#### Dereplication of MOTUs with the same scientific name
```{r results='hide'}
# Check if we have duplicated scientific_name
table(duplicated(FM_BR$scientific_name))

# 3875  taxa names are duplicates
  
# Dereplication by scientific_name
FM_BR_derep <- FM_BR %>% 
  group_by(scientific_name) %>% 
  summarise_if(is.numeric, sum)

# remove last row from data frame
FM_BR_derep <- FM_BR_derep %>% filter(row_number() <= n()-1)

# replace all NA values with zero
FM_BR_derep <- FM_BR_derep %>% replace(is.na(.), 0)
```


#### Resolving PCR replicates
```{r}
# We exclude first column (scientific_name), transpose and convert the data matrix in dataframe
FM_BR_derep_R <- data.frame(t(FM_BR_derep[,-1]))
# AND rename the table by scientific name
names(FM_BR_derep_R) <- FM_BR_derep$scientific_name

# make sample names one of the columns of the data
FM_BR_derep_R <- FM_BR_derep_R %>% 
  rownames_to_column("Sequencing_ID")

# Prepare an id that represents all three replicates (by same name) and place that after "Sequencing_ID"
FM_BR_derep_R <- FM_BR_derep_R %>% 
  add_column(Sample_ID = str_replace_all(FM_BR_derep_R$Sequencing_ID, "_R1|_R2|_R3", ""), .after = "Sequencing_ID")

# get read sum of all three replicates
FM_BR_clean_Rsum <- FM_BR_derep_R %>% 
  select(-Sequencing_ID) %>% 
  group_by(Sample_ID) %>% 
  summarise_if(is.numeric, sum)
# We have the read data in required format (Sample X Species)

# Convert un-merged read data to 0 or 1, remove first two columns
#FM_BR_derep_PCR <- data.frame(ifelse (FM_BR_derep_R[,-c(1:2)]>0, 1, 0))

# Combine with the first two columns of the previous data
#FM_BR_derep_PCR <- cbind (FM_BR_derep_R[,c(1:2)], FM_BR_derep_PCR)

# Follow the steps used while generating read sums
#FM_BR_clean_PCR <- FM_BR_derep_PCR %>% 
#  select(-Sequencing_ID) %>% 
#  group_by(Sample_ID) %>% 
#  summarise_if(is.numeric, sum)
#FM_BR_clean_PCR <- FM_BR_clean_PCR[-1,]
# We have the frequency data in required format (Sample X Species)
```


#### Identifying contaminants in marker-gene data

Creating metadata dataframe with a Sample_or_Control column for running decontam function
```{r}
FMFinn_BR_meta <- as.data.frame(FM_BR_clean_Rsum$Sample_ID)
names(FMFinn_BR_meta) <- "Sample_ID"

FMFinn_BR_meta$Sample_or_Control <- "True Sample"
FMFinn_BR_meta$Sample_or_Control[grepl("NC", FMFinn_BR_meta$Sample_ID)] <- "Control Sample"

FMFinn_BR_meta <- FMFinn_BR_meta[-1,]
```

Creating phyloseq object
```{r}
BR_to_ps <- as.data.frame(FM_BR_clean_Rsum[-1,])

# first column to rowname
row.names(BR_to_ps) <- BR_to_ps$Sample_ID
BR_to_ps <- BR_to_ps[,-1]

ps_BR <- phyloseq(otu_table(BR_to_ps, taxa_are_rows=T))
```

Identify Contaminants - with Prevalence method
```{r}
neg2 <- FMFinn_BR_meta$Sample_or_Control == 'Control Sample' # True if a negative control sample

# Prevalence-based contaminant classification
contam2 <- isContaminant(as.matrix(ps_BR), neg=neg2, threshold=0.1, detailed=TRUE, normalize=TRUE, method='prevalence')

table(contam2$contaminant)
```

Erasing from EU dataset the taxa identifed as contaminants
```{r}
# transposing
tFMFinn_BR_clean <- as.data.frame(t(FM_BR_clean_Rsum))
tFMFinn_BR_clean <- tFMFinn_BR_clean %>% row_to_names(row_number = 1)
tFMFinn_BR_clean <- tibble::rownames_to_column(tFMFinn_BR_clean, "Taxa")

# no contaminants
FMFinn_BR_decontam <- tFMFinn_BR_clean
```


#### Removing rare reads
```{r}
# Erasing negative controls
FMFinn_BR_decontam <- FMFinn_BR_decontam %>% 
       select(-contains('NC'))

# convert from character to numeric
Taxa <- as.data.frame(FMFinn_BR_decontam$Taxa)
FMFinn_BR_decontam <- as.data.frame(apply(FMFinn_BR_decontam[,-1], 2, as.numeric))
FMFinn_BR_decontam <- cbind(Taxa, FMFinn_BR_decontam)
names(FMFinn_BR_decontam)[1] = "Taxa"

# creating a count column with the sum of the readings
rf <- FMFinn_BR_decontam %>% rowwise(Taxa)
FMFinn_BR_decontam <- rf %>% mutate(count = sum(c_across(where(is.numeric))))

# ordering data by decreasing number of counts
order(FMFinn_BR_decontam$count, decreasing = T)
FMFinn_BR_decontam <- FMFinn_BR_decontam[order(FMFinn_BR_decontam$count, decreasing = T),]

# keeping only first 50 most abundant MOTUs
FMFinn_BR_50 <- FMFinn_BR_decontam[1:50, ]
```


#### Exporting dataframes

Changing format
```{r}
# transposing
tFMFinn_BR_50 <- as.data.frame(t(FMFinn_BR_50))
tFMFinn_BR_50 <- tFMFinn_BR_50 %>% 
  row_to_names(row_number = 1) 
FMFinn_BR_final <- tFMFinn_BR_50[-41,] 
#FMFinn_BR_final <- tibble::rownames_to_column(FMFinn_BR_final, "Sample_ID")
```

Exporting dataframes
```{r}
# exporting the dataframe of our samples in csv format
write.csv(FMFinn_BR_final, "outputs/FMFinn_BR_final.csv", row.names = T)
write.csv(FMFinn_BR_50, "outputs/FMFinn_BR_50.csv", row.names = F)
```



## Fungi data

#### Importing data
```{r}
# importing .tab files (tab separated files)
F_FU <- read.table(file = "data/F_FU.tab", header = T, sep = "\t")
M_FU <- read.table(file = "data/M_FU.tab", header = T, sep = "\t")
Finn_FU <- read.table(file = "data/Finn_FU.tab", header = T, sep = "\t")
```

#### Erasing "obiclean" columns
```{r}
# selecting columns not starting with "obi" 
F_FU <- F_FU %>% 
  select(-(starts_with('obi')))
M_FU <- M_FU %>% 
  select(-(starts_with('obi')))
Finn_FU <- Finn_FU %>% 
  select(-(starts_with('obi')))
```

#### Excluding MOTUs defined only as Eukaryota (superkingdom level) 
```{r}
# subsetting dataframe: all rows which scintific name is different than Eukaryota
F_FU <- F_FU %>% 
  subset(scientific_name!="Eukaryota")
M_FU <- M_FU %>% 
  subset(scientific_name!="Eukaryota")
Finn_FU <- Finn_FU %>% 
  subset(scientific_name!="Eukaryota")
```

#### Filtering data that has less than 97% identity with the Fungi reference dataset
```{r}
# subsetting dataframe: all rows which best_identity.db_FUNGI is greater than 0.97
F_FU <- F_FU %>% 
  subset(best_identity.db_FUNGI>0.97)
M_FU <- M_FU %>% 
  subset(best_identity.db_FUNGI>0.97)
Finn_FU <- Finn_FU %>% 
  subset(best_identity.db_FUNGI>0.97)
```

#### Erasing not informative data
```{r}
# erasing data that have rank of kingdom
F_FU <- F_FU %>% 
  subset(rank!="kingdom")
M_FU <- M_FU %>% 
  subset(rank!="kingdom")
Finn_FU <- Finn_FU %>% 
  subset(rank!="kingdom")
```

#### Subsetting and merging datasets for Fungi
```{r}
# selecting only column from M_FU dataset with sequence and our samples
M_FU_join <- M_FU %>% select(sequence, sample.REINS1P2Q2_S413_R1, sample.REINS1P2Q2_S413_R3, sample.REINS1P2Q2_S413_R2, sample.REINS1P2Q3_S415_R2, sample.REINS1P2Q3_S415_R1, sample.REINS1P2Q3_S415_R3, sample.EXTRNC_168_R1, sample.EXTRNC_168_R2, sample.EXTRNC_168_R3, sample.NC210618_R1, sample.NC210618_R2, sample.NC_MIX_137_R1, sample.NC_MIX_137_R2, sample.NC_MIX_137_R3, sample.NC_MIX_R1, sample.NC_MIX_R3, sample.PCRNC_MIX_1_R3, sample.PCRNC_MIX_2_R2)

# selecting only column from F_FU dataset with sequence, scientific name and our samples
F_FU_join <- F_FU %>% select(sequence, scientific_name, sample.MS1P1Q2_SH135_R1, sample.MS1P1Q2_SH135_R2, sample.MS1P1Q2_SH135_R3, sample.MS1P2Q2_SH164_R2, sample.MS1P2Q2_SH164_R3 , sample.MS1P2Q4_SH134_R1, sample.MS1P2Q4_SH134_R2, sample.MS1P2Q4_SH134_R3, sample.MS1P2Q4_SH136_R1, sample.MS1P2Q4_SH136_R2, sample.MS1P2Q4_SH136_R3, sample.MS1P3Q1_SH132_R1, sample.MS1P3Q1_SH132_R2, sample.MS1P3Q1_SH132_R3, sample.MS1P3Q1_SH133_R1, sample.MS1P3Q1_SH133_R2, sample.MS1P3Q1_SH133_R3, sample.MS2P1Q1_SH195_R1, sample.MS2P1Q1_SH195_R2, sample.MS2P1Q1_SH195_R3, sample.MS2P2Q1_SH196_R1, sample.MS2P2Q1_SH196_R2, sample.MS2P2Q1_SH196_R3, sample.HS1P1Q4_SH137_R1, sample.HS1P1Q4_SH137_R2, sample.HS1P1Q4_SH137_R3, sample.HS2P1Q3_SH193_R1, sample.HS2P1Q3_SH193_R2, sample.HS2P1Q3_SH193_R3, sample.HS2P2Q2_SH194_R1, sample.HS2P2Q2_SH194_R2, sample.HS2P2Q2_SH194_R3, sample.HS2P2Q3_SH138_R1, sample.HS2P2Q3_SH138_R2, sample.HS2P2Q3_SH138_R3, sample.HS2P2Q3_SH190_R1, sample.HS2P2Q3_SH190_R2, sample.HS2P2Q3_SH190_R3, sample.HS2P2Q3_SH192_R1, sample.HS2P2Q3_SH192_R2, sample.HS2P2Q3_SH192_R3, sample.HS2P3Q3_SH140_R1, sample.HS2P3Q3_SH140_R2, sample.HS2P3Q3_SH140_R3, sample.HS2P3Q4_SH139_R1, sample.HS2P3Q4_SH139_R2, sample.HS2P3Q4_SH139_R3, sample.PS2P1Q3_SH80_R1, sample.PS2P1Q3_SH80_R2, sample.PS2P1Q3_SH80_R3, sample.PS2P2Q2_SH100_R1, sample.PS2P2Q2_SH100_R2, sample.PS2P2Q2_SH100_R3, sample.PS2P2Q2_SH186_R1, sample.PS2P2Q2_SH186_R2, sample.PS2P2Q2_SH186_R3, sample.PS2P2Q2_SH81_R1, sample.PS2P2Q2_SH81_R2, sample.PS2P2Q2_SH81_R3, sample.PS2P2Q3_SH115_R1, sample.PS2P2Q3_SH115_R2, sample.PS2P2Q3_SH115_R3, sample.PS2P2Q4_SH113_R1, sample.PS2P2Q4_SH113_R2, sample.PS2P2Q4_SH113_R3, sample.PS2P2Q4_SH98_R1, sample.PS2P2Q4_SH98_R2, sample.PS2P2Q4_SH98_R3, sample.PS2P3Q3_SH99_R1, sample.PS2P3Q3_SH99_R2, sample.PS2P3Q3_SH99_R3, sample.RS1P2Q2_SH124_R1,  sample.RS1P2Q2_SH124_R2,  sample.RS1P2Q2_SH124_R3, sample.RS1P2Q1_SH106_R1, sample.RS1P2Q1_SH106_R2, sample.RS1P2Q1_SH106_R3, sample.RS2P1Q1_SH129_R1, sample.RS2P1Q1_SH129_R2, sample.RS2P1Q1_SH129_R3, sample.RS2P1Q2_SH93_R1, sample.RS2P1Q2_SH93_R2, sample.RS2P1Q2_SH93_R3, sample.RS2P2Q3_SH127_R1, sample.RS2P2Q3_SH127_R2, sample.RS2P2Q3_SH127_R3, sample.RS2P2Q4_SH130_R1, sample.RS2P2Q4_SH130_R2, sample.RS2P2Q4_SH130_R3, sample.RS2P2Q4_SH172_R1, sample.RS2P2Q4_SH172_R2, sample.RS2P2Q4_SH172_R3, sample.RS2P2Q4_SH176_R1, sample.RS2P2Q4_SH176_R2, sample.RS2P2Q4_SH176_R3, sample.REINS1P1Q5_S412_R1, sample.REINS1P1Q5_S412_R2, sample.REINS1P1Q5_S412_R3, sample.REINS1P2Q45_S414_R1, sample.REINS1P2Q45_S414_R3, sample.NC2_FIN_R1, sample.NC2_FIN_R2, sample.NC2_FIN_R3, sample.NC3_FIN_R1, sample.NC3_FIN_R2, sample.NC3_FIN_R3, sample.NC5_FIN_R1, sample.NC5_FIN_R2, sample.NC5_FIN_R3, sample.NC6_FIN_R1, sample.NC6_FIN_R2, sample.NC6_FIN_R3, sample.NC_FIN_R1, sample.NC_FIN_R2, sample.NC_FIN_R3, sample.PCRNC3_FIN_R1, sample.PCRNC3_FIN_R2)

# selecting only column from Finn_FU dataset with sequence and our samples
Finn_FU_join <- Finn_FU %>% select(sequence, sample.S1562_R1, sample.S1562_R2, sample.S1562_R3, sample.S1564_R1, sample.S1564_R2, sample.S1564_R3, sample.S1568_R1, sample.S1568_R2, sample.S1568_R3, sample.S1572_R1, sample.S1572_R2, sample.S1572_R3, sample.DNANCF12_R1, sample.DNANCF12_R2, sample.DNANCF12_R3, sample.DNANCF5_R1, sample.DNANCF5_R2, sample.DNANCF5_R3, sample.DNANCF8_R1, sample.DNANCF8_R2, sample.DNANCF8_R3, sample.PCRNCF4_R1, sample.PCRNCF4_R2)

# merging F_EU and M_EU in one single FM_EU dataset
FM_FU <- full_join(F_FU_join, M_FU_join, by = "sequence", copy = FALSE, keep = NULL)
FM_FU <- full_join(FM_FU, Finn_FU_join, by = "sequence", copy = FALSE, keep = NULL)
```

#### Dereplication of MOTUs with the same scientific name
```{r results='hide'}
# Check if we have duplicated scientific_name
table(duplicated(FM_FU$scientific_name))

# 16944 taxa names are duplicates
  
# Dereplication by scientific_name
FM_FU_derep <- FM_FU %>% 
  group_by(scientific_name) %>% 
  summarise_if(is.numeric, sum)

# remove last row from data frame
FM_FU_derep <- FM_FU_derep %>% filter(row_number() <= n()-1)

# replace all NA values with zero
FM_FU_derep <- FM_FU_derep %>% replace(is.na(.), 0)
```


#### Resolving PCR replicates
```{r}
# We exclude first column (scientific_name), transpose and convert the data matrix in dataframe
FM_FU_derep_R <- data.frame(t(FM_FU_derep[,-1]))
# AND rename the table by scientific name
names(FM_FU_derep_R) <- FM_FU_derep$scientific_name

# make sample names one of the columns of the data
FM_FU_derep_R <- FM_FU_derep_R %>% 
  rownames_to_column("Sequencing_ID")

# Prepare an id that represents all three replicates (by same name) and place that after "Sequencing_ID"
FM_FU_derep_R <- FM_FU_derep_R %>% 
  add_column(Sample_ID = str_replace_all(FM_FU_derep_R$Sequencing_ID, "_R1|_R2|_R3", ""), .after = "Sequencing_ID")

# get read sum of all three replicates
FM_FU_clean_Rsum <- FM_FU_derep_R %>% 
  select(-Sequencing_ID) %>% 
  group_by(Sample_ID) %>% 
  summarise_if(is.numeric, sum)
# We have the read data in required format (Sample X Species)

# Convert un-merged read data to 0 or 1, remove first two columns
#FM_FU_derep_PCR <- data.frame(ifelse (FM_FU_derep_R[,-c(1:2)]>0, 1, 0))

# Combine with the first two columns of the previous data
#FM_FU_derep_PCR <- cbind(FM_FU_derep_R[,c(1:2)], FM_FU_derep_PCR)

# Follow the steps used while generating read sums
#FM_FU_clean_PCR <- FM_FU_derep_PCR %>% 
#  select(-Sequencing_ID) %>% 
#  group_by(Sample_ID) %>% 
#  summarise_if(is.numeric, sum)
#FM_FU_clean_PCR <- FM_FU_clean_PCR[-1,]
# We have the frequency data in required format (Sample X Species)
```


#### Identifying contaminants in marker-gene data

Creating FMFinn_EU metadata dataframe with a Sample_or_Control column for running decontam function
```{r}
FMFinn_FU_meta <- as.data.frame(FM_FU_clean_Rsum$Sample_ID)
names(FMFinn_FU_meta) <- "Sample_ID"

FMFinn_FU_meta$Sample_or_Control <- "True Sample"
FMFinn_FU_meta$Sample_or_Control[grepl("NC", FMFinn_FU_meta$Sample_ID)] <- "Control Sample"

FMFinn_FU_meta <- FMFinn_FU_meta[-1,]
```

Creating phyloseq object
```{r}
FU_to_ps <- as.data.frame(FM_FU_clean_Rsum[-1,])

# first column to rowname
row.names(FU_to_ps) <- FU_to_ps$Sample_ID
FU_to_ps <- FU_to_ps[,-1]

ps_FU <- phyloseq(otu_table(FU_to_ps, taxa_are_rows=T))
```

Identify Contaminants - with Prevalence method
```{r}
neg3 <- FMFinn_FU_meta$Sample_or_Control == 'Control Sample' # True if a negative control sample

# Prevalence-based contaminant classification
contam3 <- isContaminant(as.matrix(ps_FU), neg=neg3, threshold=0.1, detailed=TRUE, normalize=TRUE, method='prevalence')

table(contam3$contaminant)
```

Erasing from FU dataset the taxa identifed as contaminants
```{r}
# right format for df to merge
tFMFinn_FU_clean <- as.data.frame(t(FM_FU_clean_Rsum))
tFMFinn_FU_clean <- tFMFinn_FU_clean %>% row_to_names(row_number = 1)
tFMFinn_FU_clean <- tibble::rownames_to_column(tFMFinn_FU_clean, "Taxa")

contam3 <- tibble::rownames_to_column(contam3, "Taxa")

contaminant3 <- contam3 %>%
   dplyr::select(Taxa, contaminant)

# merge contaminant column with EU dataset
FMFinn_FU_contam <- left_join(tFMFinn_FU_clean, contaminant3, by = "Taxa")

# subset only non contaminants
FMFinn_FU_decontam <- FMFinn_FU_contam %>%
  subset(contaminant == "FALSE") %>%
  dplyr::select(-contaminant)
```


#### Removing rare reads
We need to decide if for the purposes of our project we want to select a treshold which is a percentage of the dataset (eg. 1 per thousand most abundant MOTUs) or an arbitrary number of MOTUs considered as predominant in the diet (eg. 50 most abundant MOTUs).
```{r}
# Erasing negative controls
FMFinn_FU_decontam <- FMFinn_FU_decontam %>% 
       select(-contains('NC'))

# convert from character to numeric
Taxa <- as.data.frame(FMFinn_FU_decontam$Taxa)
FMFinn_FU_decontam <- as.data.frame(apply(FMFinn_FU_decontam[,-1], 2, as.numeric))
FMFinn_FU_decontam <- cbind(Taxa, FMFinn_FU_decontam)
names(FMFinn_FU_decontam)[1] = "Taxa"

# creating a count column with the sum of the readings
rf <- FMFinn_FU_decontam %>% rowwise(Taxa)
FMFinn_FU_decontam <- rf %>% mutate(count = sum(c_across(where(is.numeric))))

# ordering data by decreasing number of counts
order(FMFinn_FU_decontam$count, decreasing = T)
FMFinn_FU_decontam <- FMFinn_FU_decontam[order(FMFinn_FU_decontam$count, decreasing = T),]

# keeping only first 50 most abundant MOTUs
FMFinn_FU_50 <- FMFinn_FU_decontam[1:50, ]
```


#### Exporting dataframes

Changing format
```{r}
# transposing
tFMFinn_FU_50 <- as.data.frame(t(FMFinn_FU_50))
tFMFinn_FU_50 <- tFMFinn_FU_50 %>% 
  row_to_names(row_number = 1) 
FMFinn_FU_final <- tFMFinn_FU_50[-41,] 
#FMFinn_FU_final <- tibble::rownames_to_column(FMFinn_FU_final, "Sample_ID")
```

Exporting dataframes
```{r}
# exporting the dataframe of our samples in csv format
write.csv(FMFinn_FU_final, "outputs/FMFinn_FU_final.csv", row.names = T)
write.csv(FMFinn_FU_50, "outputs/FMFinn_FU_50.csv", row.names = F)
```



## Plant data

#### Importing data
```{r}
# importing .tab files (tab separated files)
F_GH <- read.table(file = "data/F_GH.tab", header = T, sep = "\t")
M_GH <- read.table(file = "data/M_GH.tab", header = T, sep = "\t")
Finn_GH <- read.table(file = "data/Finn_GH.tab", header = T, sep = "\t")
```

#### Erasing "obiclean" columns
```{r}
# selecting columns not starting with "obi" 
F_GH <- F_GH %>% 
  select(-(starts_with('obi')))
M_GH <- M_GH %>% 
  select(-(starts_with('obi')))
Finn_GH <- Finn_GH %>% 
  select(-(starts_with('obi')))
```

#### Excluding MOTUs defined only as Eukaryota (superkingdom level) 
```{r}
# subsetting dataframe: all rows which scientific name is different than Eukaryota
F_GH <- F_GH %>% 
  subset(scientific_name!="Eukaryota")
M_GH <- M_GH %>% 
  subset(scientific_name!="Eukaryota")
Finn_GH <- Finn_GH %>% 
  subset(scientific_name!="Eukaryota")
```

#### Filtering data that has less than 97% identity with the Plant reference dataset
```{r}
# subsetting dataframe: all rows which best_identity.arctborbryo.gh is greater than 0.97
F_GH <- F_GH %>% 
  subset(best_identity.arctborbryo.gh>0.97)
M_GH <- M_GH %>% 
  subset(best_identity.arctborbryo.gh>0.97)
Finn_GH <- Finn_GH %>% 
  subset(best_identity.db_GH>0.97)
```

#### Erasing not informative data
```{r}
# erasing data that have rank of kingdom
F_GH <- F_GH %>% 
  subset(rank!="kingdom")
M_GH <- M_GH %>% 
  subset(rank!="kingdom")
Finn_GH <- Finn_GH %>% 
  subset(rank!="kingdom")
```

#### Subsetting and merging datasets for Plants
```{r}
# selecting only column from M_GH dataset with sequence and our samples
M_GH_join <- M_GH %>% select(sequence, sample.REINS1P2Q2_S413_R1, sample.REINS1P2Q2_S413_R2, sample.REINS1P2Q2_S413_R3, sample.REINS1P2Q3_S415_R1, sample.REINS1P2Q3_S415_R3, sample.DNANC_140_R1, sample.DNANC_140_R2, sample.DNANC_140_R3, sample.NC210618_R1, sample.NC210618_R2, sample.NC210618_R3, sample.EXTRNC_168_R1, sample.EXTRNC_168_R2, sample.EXTRNC_168_R3, sample.NC_MIX_137_R1, sample.NC_MIX_137_R2, sample.NC_MIX_137_R3, sample.PCRNC_MIX_2_R1, sample.PCRNC_MIX_2_R2, sample.PCRNC_MIX_2_R3)

# selecting only column from F_GH dataset with sequence, scientific name and our samples
F_GH_join <- F_GH %>% select(sequence, scientific_name, sample.MS1P1Q2_SH135_R1, sample.MS1P1Q2_SH135_R2, sample.MS1P1Q2_SH135_R3, sample.MS1P2Q2_SH164_R2, sample.MS1P2Q2_SH164_R3 , sample.MS1P2Q4_SH134_R1, sample.MS1P2Q4_SH134_R2, sample.MS1P2Q4_SH134_R3, sample.MS1P2Q4_SH136_R1, sample.MS1P2Q4_SH136_R2, sample.MS1P2Q4_SH136_R3, sample.MS1P3Q1_SH132_R1, sample.MS1P3Q1_SH132_R2, sample.MS1P3Q1_SH132_R3, sample.MS1P3Q1_SH133_R1, sample.MS1P3Q1_SH133_R2, sample.MS1P3Q1_SH133_R3, sample.MS2P1Q1_SH195_R1, sample.MS2P1Q1_SH195_R2, sample.MS2P1Q1_SH195_R3, sample.MS2P2Q1_SH196_R1, sample.MS2P2Q1_SH196_R2, sample.MS2P2Q1_SH196_R3, sample.HS1P1Q4_SH137_R1, sample.HS1P1Q4_SH137_R2, sample.HS1P1Q4_SH137_R3, sample.HS2P1Q3_SH193_R1, sample.HS2P1Q3_SH193_R2, sample.HS2P1Q3_SH193_R3, sample.HS2P2Q2_SH194_R1, sample.HS2P2Q2_SH194_R2, sample.HS2P2Q2_SH194_R3, sample.HS2P2Q3_SH138_R1, sample.HS2P2Q3_SH138_R2, sample.HS2P2Q3_SH138_R3, sample.HS2P2Q3_SH190_R1, sample.HS2P2Q3_SH190_R2, sample.HS2P2Q3_SH190_R3, sample.HS2P2Q3_SH192_R1, sample.HS2P2Q3_SH192_R2, sample.HS2P2Q3_SH192_R3, sample.HS2P3Q3_SH140_R1, sample.HS2P3Q3_SH140_R2, sample.HS2P3Q3_SH140_R3, sample.HS2P3Q4_SH139_R1, sample.HS2P3Q4_SH139_R2, sample.HS2P3Q4_SH139_R3, sample.PS2P1Q3_SH80_R1, sample.PS2P1Q3_SH80_R2, sample.PS2P1Q3_SH80_R3, sample.PS2P2Q2_SH100_R1, sample.PS2P2Q2_SH100_R2, sample.PS2P2Q2_SH100_R3, sample.PS2P2Q2_SH186_R1, sample.PS2P2Q2_SH186_R2, sample.PS2P2Q2_SH186_R3, sample.PS2P2Q2_SH81_R1, sample.PS2P2Q2_SH81_R2, sample.PS2P2Q2_SH81_R3, sample.PS2P2Q3_SH115_R1, sample.PS2P2Q3_SH115_R2, sample.PS2P2Q3_SH115_R3, sample.PS2P2Q4_SH113_R1, sample.PS2P2Q4_SH113_R2, sample.PS2P2Q4_SH113_R3, sample.PS2P2Q4_SH98_R1, sample.PS2P2Q4_SH98_R2, sample.PS2P2Q4_SH98_R3, sample.PS2P3Q3_SH99_R1, sample.PS2P3Q3_SH99_R2, sample.PS2P3Q3_SH99_R3, sample.RS1P2Q2_SH124_R1,  sample.RS1P2Q2_SH124_R2,  sample.RS1P2Q2_SH124_R3, sample.RS1P2Q1_SH106_R1, sample.RS1P2Q1_SH106_R2, sample.RS1P2Q1_SH106_R3, sample.RS2P1Q1_SH129_R1, sample.RS2P1Q1_SH129_R2, sample.RS2P1Q1_SH129_R3, sample.RS2P1Q2_SH93_R1, sample.RS2P1Q2_SH93_R2, sample.RS2P1Q2_SH93_R3, sample.RS2P2Q3_SH127_R1, sample.RS2P2Q3_SH127_R2, sample.RS2P2Q3_SH127_R3, sample.RS2P2Q4_SH130_R1, sample.RS2P2Q4_SH130_R2, sample.RS2P2Q4_SH130_R3, sample.RS2P2Q4_SH172_R1, sample.RS2P2Q4_SH172_R2, sample.RS2P2Q4_SH172_R3, sample.RS2P2Q4_SH176_R1, sample.RS2P2Q4_SH176_R2, sample.RS2P2Q4_SH176_R3, sample.REINS1P1Q5_S412_R1, sample.REINS1P1Q5_S412_R2, sample.REINS1P1Q5_S412_R3, sample.REINS1P2Q45_S414_R1, sample.REINS1P2Q45_S414_R3, sample.NC2_FIN_R1, sample.NC2_FIN_R2, sample.NC2_FIN_R3, sample.NC5_FIN_R1, sample.NC5_FIN_R2, sample.NC5_FIN_R3, sample.OANC2_R1, sample.OANC2_R2, sample.OANC2_R3, sample.NC3_FIN_R1, sample.NC4_FIN_R1, sample.NC_FIN_R1, sample.OANC_R1)

# selecting only column from Finn_GH dataset with sequence and our samples
Finn_GH_join <- Finn_GH %>% select(sequence, sample.S1562_R1, sample.S1562_R2, sample.S1562_R3, sample.S1564_R1, sample.S1564_R2, sample.S1564_R3, sample.S1568_R1, sample.S1568_R2, sample.S1568_R3, sample.S1572_R1, sample.S1572_R2, sample.S1572_R3, sample.DNANCF12_R2, sample.DNANCF12_R2, sample.DNANCF2_R2, sample.DNANCF2_R2, sample.DNANCF5_R2, sample.DNANCF5_R3, sample.PCRNCF1_R3, sample.PCRNCF3_R1, sample.PCRNCF5_R1, sample.PCRNCF8_R1)

# merging F_BR and M_BR in one single FM_BR dataset
FM_GH <- full_join(F_GH_join, M_GH_join, by = "sequence", copy = FALSE, keep = NULL)
FMFinn_GH <- full_join(FM_GH, Finn_GH_join, by = "sequence", copy = FALSE, keep = NULL)
```

#### Dereplication of MOTUs with the same scientific name
```{r results='hide'}
# Check if we have duplicated scientific_name
table(duplicated(FMFinn_GH$scientific_name))

# 7050 taxa names are duplicates
  
# Dereplication by scientific_name
FMFinn_GH_derep <- FMFinn_GH %>% 
  group_by(scientific_name) %>% 
  summarise_if(is.numeric, sum)

# remove last row from data frame
FMFinn_GH_derep <- FMFinn_GH_derep %>% filter(row_number() <= n()-1)

# replace all NA values with zero
FMFinn_GH_derep <- FMFinn_GH_derep %>% replace(is.na(.), 0)
```


#### Resolving PCR replicates
```{r}
# We exclude first column (scientific_name), transpose and convert the data matrix in dataframe
FMFinn_GH_derep_R <- data.frame(t(FMFinn_GH_derep[,-1]))
# AND rename the table by scientific name
names(FMFinn_GH_derep_R) <- FMFinn_GH_derep$scientific_name

# make sample names one of the columns of the data
FMFinn_GH_derep_R <- FMFinn_GH_derep_R %>% 
  rownames_to_column("Sequencing_ID")

# Prepare an id that represents all three replicates (by same name) and place that after "Sequencing_ID"
FMFinn_GH_derep_R <- FMFinn_GH_derep_R %>% 
  add_column(Sample_ID = str_replace_all(FMFinn_GH_derep_R$Sequencing_ID, "_R1|_R2|_R3", ""), .after = "Sequencing_ID")

# get read sum of all three replicates
FMFinn_GH_clean_Rsum <- FMFinn_GH_derep_R %>% 
  select(-Sequencing_ID) %>% 
  group_by(Sample_ID) %>% 
  summarise_if(is.numeric, sum)
# We have the read data in required format (Sample X Species)

# Convert un-merged read data to 0 or 1, remove first two columns
#FMFinn_GH_derep_PCR <- data.frame(ifelse (FMFinn_GH_derep_R[,-c(1:2)]>0, 1, 0))

# Combine with the first two columns of the previous data
#FMFinn_GH_derep_PCR <- cbind(FMFinn_GH_derep_R[,c(1:2)], FMFinn_GH_derep_PCR)

# Follow the steps used while generating read sums
#FMFinn_GH_clean_PCR <- FMFinn_GH_derep_PCR %>% 
#  select(-Sequencing_ID) %>% 
#  group_by(Sample_ID) %>% 
#  summarise_if(is.numeric, sum)
#FMFinn_GH_clean_PCR <- FMFinn_GH_clean_PCR[-1,]
# We have the frequency data in required format (Sample X Species)
```


#### Identifying contaminants in marker-gene data

Creating metadata dataframe with a Sample_or_Control column for running decontam function
```{r}
FMFinn_GH_meta <- as.data.frame(FMFinn_GH_clean_Rsum$Sample_ID)
names(FMFinn_GH_meta) <- "Sample_ID"

FMFinn_GH_meta$Sample_or_Control <- "True Sample"
FMFinn_GH_meta$Sample_or_Control[grepl("NC", FMFinn_GH_meta$Sample_ID)] <- "Control Sample"

FMFinn_GH_meta <- FMFinn_GH_meta[-1,]
```

Creating phyloseq object
```{r}
GH_to_ps <- as.data.frame(FMFinn_GH_clean_Rsum[-1,])

# first column to rowname
row.names(GH_to_ps) <- GH_to_ps$Sample_ID
GH_to_ps <- GH_to_ps[,-1]

ps_GH <- phyloseq(otu_table(GH_to_ps, taxa_are_rows=T))
```

Identify Contaminants - with Prevalence method
```{r}
neg4 <- FMFinn_GH_meta$Sample_or_Control == 'Control Sample' # True if a negative control sample

# Prevalence-based contaminant classification
contam4 <- isContaminant(as.matrix(ps_GH), neg=neg4, threshold=0.1, detailed=TRUE, normalize=TRUE, method='prevalence')

table(contam4$contaminant)
```

Erasing from EU dataset the taxa identifed as contaminants
```{r}
# right format for df to merge
tFMFinn_GH_clean <- as.data.frame(t(FMFinn_GH_clean_Rsum))
tFMFinn_GH_clean <- tFMFinn_GH_clean %>% row_to_names(row_number = 1)
tFMFinn_GH_clean <- tibble::rownames_to_column(tFMFinn_GH_clean, "Taxa")

contam4 <- tibble::rownames_to_column(contam4, "Taxa")

contaminant4 <- contam4 %>%
   dplyr::select(Taxa, contaminant)

# merge contaminant column with EU dataset
FMFinn_GH_contam <- left_join(tFMFinn_GH_clean, contaminant4, by = "Taxa")

# subset only non contaminants
FMFinn_GH_decontam <- FMFinn_GH_contam %>%
  subset(contaminant == "FALSE") %>%
  dplyr::select(-contaminant)
```


#### Removing rare reads
```{r}
# Erasing negative controls
FMFinn_GH_decontam <- FMFinn_GH_decontam %>% 
       select(-contains('NC'))

# convert from character to numeric
Taxa <- as.data.frame(FMFinn_GH_decontam$Taxa)
FMFinn_GH_decontam <- as.data.frame(apply(FMFinn_GH_decontam[,-1], 2, as.numeric))
FMFinn_GH_decontam <- cbind(Taxa, FMFinn_GH_decontam)
names(FMFinn_GH_decontam)[1] = "Taxa"

# creating a count column with the sum of the readings
rf <- FMFinn_GH_decontam %>% rowwise(Taxa)
FMFinn_GH_decontam <- rf %>% mutate(count = sum(c_across(where(is.numeric))))

# ordering data by decreasing number of counts
order(FMFinn_GH_decontam$count, decreasing = T)
FMFinn_GH_decontam <- FMFinn_GH_decontam[order(FMFinn_GH_decontam$count, decreasing = T),]

# keeping only first 50 most abundant MOTUs
FMFinn_GH_50 <- FMFinn_GH_decontam[1:50, ]
```


#### Exporting dataframes

Changing format
```{r}
# transposing
tFMFinn_GH_50 <- as.data.frame(t(FMFinn_GH_50))
tFMFinn_GH_50 <- tFMFinn_GH_50 %>% 
  row_to_names(row_number = 1) 
FMFinn_GH_final <- tFMFinn_GH_50[-41,] 
#FMFinn_GH_final <- tibble::rownames_to_column(FMFinn_GH_final, "Sample_ID")
```

Exporting dataframes
```{r}
# exporting the dataframe of our samples in csv format
write.csv(FMFinn_GH_final, "outputs/FMFinn_GH_final.csv", row.names = T)
write.csv(FMFinn_GH_50, "outputs/FMFinn_GH_50.csv", row.names = F)
```


#### Creating categories: plants
```{r}
# import csv file with Categories
GH_Cat <- read.csv("data/plant_cat.csv", header = T, sep = ";")

# order levels of factor
GH_Cat$category <- factor(GH_Cat$category, 
                        levels=c("Betula","Fern","Forb","Graminoid", "Moss", "Salicacea", "Shrub", "Tree","Vaccinium"))
GH_Cat$category <- factor(GH_Cat$category, levels = rev(levels(GH_Cat$category)))

# substitute spaces with underscores
GH_Cat$taxa <- gsub(pattern = "\\ ", replacement="_", 
          x=GH_Cat$taxa )
```



